To add script 01A for the upstream-analysis module (#20)
stjude-dnb-binfcore/sc-epigenie: analyses/upstream-analysis/01A_run_signac_qc.Rmd

source(paste0(analysis_dir, "/util/function-process-signac.R"))
source(paste0(analysis_dir, "/util/function-create-UMAP.R"))
```

```{r echo=FALSE, warning=FALSE}
opts_chunk$set(fig.align='center',
               external=TRUE,
               echo=FALSE,
               warning=FALSE,
               fig.pos='H')
a4width <- 8.3
a4height <- 11.7
```

# Read metadata file and define `sample_name`

```{r read-project-metadat, echo=TRUE}
project_metadata <- read.csv(project_metadata_file, sep = "\t", header = TRUE)
sample_name <- unique(as.character(project_metadata$ID))
sample_name <- sort(sample_name, decreasing = FALSE)
print(sample_name)
```

# Read palette file(s)

```{r define-parameters-for-plots, echo=TRUE}
# Fix random seed for reproducible results
set.seed(12345)

# Read color palette
palette_df <- readr::read_tsv(palette_file, guess_max = 100000, show_col_types = FALSE) 
gradient_palette_df <- readr::read_tsv(gradient_palette_file, guess_max = 100000, show_col_types = FALSE) 
```

# Read in 10x data 

When pre-processing chromatin data, Signac uses information from two related input files, both of which can be created using CellRanger:

> - **Peak/Cell matrix.** This is analogous to the gene expression count matrix used to analyze single-cell RNA-seq. **However, instead of genes, each row of the matrix represents a region of the genome (a peak), that is predicted to represent a region of open chromatin.** Each value in the matrix represents the number of Tn5 integration sites for each single barcode (i.e. a cell) that map within each peak. You can find more detail on the 10X Website.

Author: See bdf98e5e6caf0a998e1e493b67493b36ab76a108.

```{r read-process-data, echo=TRUE}
cat("Beginning to process sample:", sample_name[i], "\n")

# Read files
counts <- Read10X_h5(filename = counts_file)
meta <- read.csv(file = meta_file, header = TRUE, row.names = 1)

# Create chromatin and seurat objects
chrom_assay <- CreateChromatinAssay(counts = counts, sep = c(":", "-"), genome = genome_name, fragments = fragments_file, min.cells = 10, min.features = 200)
seurat_obj <- CreateSeuratObject(counts = chrom_assay, assay = assay, meta.data = meta)

# Assign a sample ID to all cells from a specific sample
seurat_obj$ID <- sample_name[i]

# Add metadata from `project_metadata` file
project_metadata_sample <- project_metadata %>% dplyr::filter(ID == sample_name[i])

# Ensure rownames in metadata match those IDs
rownames(project_metadata_sample) <- project_metadata_sample$ID

# Merge using AddMetaData()
seurat_obj <- AddMetaData(seurat_obj, metadata = project_metadata_sample[seurat_obj$ID, ])

# Update orig.ident using ID column
seurat_obj$orig.ident <- seurat_obj@meta.data$ID

cat("Number of cells before filter for", sample_name[i], ":", "\n")
cells_before_filter_num <- print(ncol(seurat_obj)) # Number of cells before filter
```

> 

Author: @sharonfreshour Can you please point me to the relevant code?

> 

Author: Hmmm, I have never seen this before. It doesn't seem to be part of [the brain vignette](https://stuartlab.org/signac/articles/mouse_brain_vignette). I wonder if this step is species/tissue specific.

> 

Author: @sharonfreshour See ab46ed64be15cc263cb59fe3ffea93fbaf080dd4.

```{r TSSPlot-before-filtering, fig.width = 6, fig.height = 5, fig.fullwidth = TRUE, echo=TRUE}
name <- paste0(plots_dir, "/", glue::glue("02-TSSPlot-before-filtering.png"))
p <- TSSPlot(seurat_obj, group.by = 'high.tss') + NoLegend()
p <- print(p + ggtitle(sample_name[i])) # Add a dynamic title using sample_name[i]
ggsave(file = name, width = 6, height = 5, device = "png")
```

Next, we can look at the fragment length periodicity for all the cells, and group by cells with high or low nucleosomal signal strength. That is known as nucleosome signal and it is a measure of how well nucleosome positioning is preserved in the ATAC-seq data. It is calculated based on the periodicity of fragment lengths (mononucleosome, dinucleosome, etc.). Example:

| Group     | Meaning                                   | Count  |
| --------- | ----------------------------------------- | ------ |
| `Low NS`  | Nucleosome signal < 4 → likely good cells | 19,972 |

> | `High NS` | Nucleosome signal > 4 → potentially noisy | 34     |

Author: See bdf98e5e6caf0a998e1e493b67493b36ab76a108.

The combined steps of TF-IDF followed by SVD are known as latent semantic indexing (LSI), and were first introduced for the analysis of scATAC-seq data by [Cusanovich et al. 2015](https://www.science.org/doi/10.1126/science.aab1601). The first LSI component often captures sequencing depth (technical variation) rather than biological variation. If this is the case, the component should be removed from downstream analysis. We can assess the correlation between each LSI component and sequencing depth using the DepthCor() function. This plot can indicate correlation between the first LSI component and the total number of counts for the cell. If there is a very strong correlation between the first LSI component and the total number of counts for the cell, then we will perform downstream steps without this component as we don’t want to group cells together based on their total sequencing depth, but rather by their patterns of accessibility at cell-type-specific peaks.

> ## Non-linear dimension reduction and clustering
> 
> Now that the cells are embedded in a low-dimensional space we can use methods commonly applied for the analysis of scRNA-seq data to perform graph-based clustering and non-linear dimension reduction for visualization. The functions RunUMAP(), FindNeighbors(), and FindClusters() all come from the Seurat package.

Reviewer: In this section, the `process_signac` function is only running the `RunUMAP()`, so perhaps this could be edited for clarity that the additional clustering steps are not being done here, only the UMAP dimension reduction.

```suggestion
## Non-linear dimension reduction

Now that the cells are embedded in a low-dimensional space we can use methods commonly applied for the analysis of scRNA-seq data to perform non-linear dimension reduction for visualization. The function RunUMAP() will be used from the Seurat package.
```

Author: See 54a495b035cbd10e47b3f78d7c7626068353d54b.
